pcie/server/pcie.c

```diff
@@ -0,0 +1,296 @@
+#define PCI_SUBORDINATE_BUS 0x1a
+#define PCI_CAP_PTR         0x34
+
+#define PCI_HT_MULTI_FUNC 0x80
+
+/* ECAM addressing */
+#define ECAM_BUS_SHIFT  20
+#define ECAM_DEV_SHIFT  15
+#define ECAM_FUNC_SHIFT 12
+
+static void pcie_scanBus(void *ecam, uint8_t bus);
+
+static inline volatile uint32_t *ecamRegPtr(void *ecam, uint8_t bus, uint8_t dev, uint8_t fn, uint16_t reg)
+{
+	uintptr_t cfg_space_offset = ((uintptr_t)bus << ECAM_BUS_SHIFT) |
+			((uintptr_t)dev << ECAM_DEV_SHIFT) |
+			((uintptr_t)fn << ECAM_FUNC_SHIFT) |
+			(reg & 0xfff);
+
+	return (volatile uint32_t *)((uintptr_t)ecam + cfg_space_offset);
+}
+
+static inline uint32_t ecamRead32(void *ecam, uint8_t bus, uint8_t dev, uint8_t fn, uint16_t off)
+{
+	return *ecamRegPtr(ecam, bus, dev, fn, off & ~0x3);
+}
+
+static inline void ecamWrite32(void *ecam, uint8_t bus, uint8_t dev, uint8_t fn, uint16_t off, uint32_t val)
+{
+	*ecamRegPtr(ecam, bus, dev, fn, off & ~0x3) = val;
+}
+
+static inline uint16_t ecamRead16(void *ecam, uint8_t bus, uint8_t dev, uint8_t fn, uint16_t off)
+{
+	uint32_t value_u32 = ecamRead32(ecam, bus, dev, fn, off);
+	if (off & 2) {
+		return value_u32 >> 16;
+	}
+	else {
+		return value_u32 & 0xffff;
+	}
+}
+
+static inline uint8_t ecamRead8(void *ecam, uint8_t bus, uint8_t dev, uint8_t fn, uint16_t off)
+{
+	uint32_t value_u32 = ecamRead32(ecam, bus, dev, fn, off);
+	return (value_u32 >> ((off & 3) * 8)) & 0xff;
+}
+
+static void print_bars(void *ecam, uint8_t bus, uint8_t dev, uint8_t fn, uint8_t hdr)
+{
+	/* Choose number of bars depending on config type */
+	const int bar_count = (hdr == 0x00) ? 6 : 2;
+
+	/* Read all bars */
+	for (int i = 0; i < bar_count; ++i) {
+
+		uint32_t bar_low = ecamRead32(ecam, bus, dev, fn, PCI_BAR0 + i * 4);
+		if (bar_low == 0) {
+			continue;
+		}
+
+		if ((bar_low & 0x1) == 0x1) {
+			printf("pcie: BAR%d I/O 0x%08x\n", i, bar_low & ~0x3);
+		}
+		else {
+			bool is_64_bit = (bar_low & 0x4) == 0x4;
+			uint64_t addr = bar_low & ~0xf;
+
+			if (is_64_bit) {
+				uint32_t bar_high = ecamRead32(ecam, bus, dev, fn, PCI_BAR0 + (i + 1) * 4);
+				addr |= ((uint64_t)bar_high) << 32;
+				i++;
+			}
+			printf("pcie: BAR%d MEM 0x%016llx (%s)\n",
+					i, (unsigned long long)addr, is_64_bit ? "64-bit" : "32-bit");
+		}
+	}
+}
+
+static void print_capabilities(void *ecam, uint8_t bus, uint8_t dev, uint8_t fn)
+{
+	/* Check if there is capabilities list */
+	uint16_t status = ecamRead16(ecam, bus, dev, fn, PCI_STATUS);
+	if (!(status & (1 << 4))) {
+		return;
+	}
+
+	/* Read capabilities list pointer */
+	uint8_t ptr = ecamRead8(ecam, bus, dev, fn, PCI_CAP_PTR);
+	while (ptr >= 0x40) {
+		uint8_t cap_id = ecamRead8(ecam, bus, dev, fn, ptr);
+		uint8_t next = ecamRead8(ecam, bus, dev, fn, ptr + 1);
+		printf("pcie: CAP id 0x%02x address 0x%02x\n", cap_id, ptr);
+		if (next == 0) {
+			break;
+		}
+		ptr = next;
+	}
+}
+
+static void scanFunc(void *ecam, uint8_t bus, uint8_t dev, uint8_t fun)
+{
+	static uint8_t next_bus = 1;
```

Comment:
Would be much better to keep this value in `pcie_scanBus` and pass it as an argument - right now this causes `pcie_scanBus` to be strictly one-time use